xds/internal/balancer/cdsbalancer/cluster_handler.go

```diff
@@ -0,0 +1,201 @@
+func createClusterNode(clusterName string, xdsClient xdsClientInterface, topLevelHandler *clusterHandler) *clusterNode {
+	print("|createClusterNode() called on ", clusterName, "|")
+	c := &clusterNode{
+		clusterHandler: topLevelHandler,
+	}
+	print("About to communicate with the xds client for:", clusterName, "|")
+	// Communicate with the xds client here.
+	c.cancelFunc = xdsClient.WatchCluster(clusterName, c.handleResp)
+	return c
+}
+
+// This function cancels the cluster watch on the cluster and all of it's children.
+func (c *clusterNode) delete() {
+	c.cancelFunc()
+	for _, child := range c.children {
+		child.delete()
+	}
+}
+
+// Construct cluster update (potentially a list of ClusterUpdates) for a node.
+func (c *clusterNode) constructClusterUpdate() ([]xdsclient.ClusterUpdate, error) { // THIS LOGIC IS WRONG
+	print("Tries constructing a cluster update for Cluster ", c.clusterUpdate.ServiceName, ". ")
+	// If the cluster has not yet received an update, the cluster update is not yet ready.
+	if !c.receivedUpdate {
+		return nil, errors.New("Tried to construct a cluster update on a cluster that has not received an update.")
+	}
+
+	// Base case - LogicalDNS or EDS. Both of these cluster types will be tied to a single ClusterUpdate.
+	if c.clusterUpdate.ClusterType != xdsclient.ClusterTypeAggregate {
+		print("Returning the cluster update for service (base node):", c.clusterUpdate.ServiceName)
+		return []xdsclient.ClusterUpdate{c.clusterUpdate}, nil
+	}
+
+	// If an aggregate construct a list by recursively calling down to all of it's children.
+	var childrenUpdates []xdsclient.ClusterUpdate
+	for _, child := range c.children {
+		childUpdateList, err := child.constructClusterUpdate()
+		if err != nil {
+			return nil, err
+		}
+		childrenUpdates = append(childrenUpdates, childUpdateList...)
+	}
+	print("Returning the update list for an aggregate cluster")
+	return childrenUpdates, nil
+}
+
+// handleResp handles a xds response for a particular cluster. This function also
+// handles any logic with regards to any child state that may have changed.
+// Problem statement: perhaps this logic is wrong
+func (c *clusterNode) handleResp(clusterUpdate xdsclient.ClusterUpdate, err error) {
+	c.clusterHandler.clusterMutex.Lock()
+	defer c.clusterHandler.clusterMutex.Unlock()
+	print("SERVICE NAME:", clusterUpdate.ServiceName)
+	if err != nil { // Write this error for run() to pick up in CDS LB policy.
+		c.clusterHandler.updateChannel<-clusterHandlerUpdate{chu: nil, err: err}
+		return
+	}
+	// This variable will determine whether there was a delta with regards to this clusterupdate. If there was, at the end
+	// of the response ping ClusterHandler to send CDS Policy a new list of ClusterUpdates.
+	var delta bool
+	// If this is the first update to the Cluster Node, ping ClusterHandler at the end of method to try and build a config.
+	if !c.receivedUpdate {
+		delta = true
+	}
+
+	c.receivedUpdate = true
+	c.clusterUpdate = clusterUpdate
```

Comment:
You need to handle the case that a leaf node receives a different update, to notify the top handler (or maybe I missed that somewhere).

And if this is not currently covered by test case, add one.

Reply:
Yup good catch. Will add test case.